### PR TITLE
eliminate locking bugs

### PR DESCRIPTION
**Pull request by DavidVorick**
there are a few locking problems in the consensus set that were causing database inconsistencies. They should be fixed now.

### CODE REVIEW

**Comment by DavidVorick** [comment]
oh I didn't realize this PR was open.

I want a chance to review the diff and make sure it's all sane.

File: modules/consensus/applytransaction.go [hunk 1]
```diff
@@ -26,13 +26,22 @@ var (
...
 	for _, sci := range t.SiacoinInputs {
+		scoBytes := scoBucket.Get(sci.ParentID[:])
+		if scoBytes == nil {
+			panic("nil sco bytes")
+		}
+		var sco types.SiacoinOutput
+		err := encoding.Unmarshal(scoBytes, &sco)
+		if err != nil {
+			return err
+		}
```
Comment (DavidVorick) [inline modules/consensus/applytransaction.go]: I am worried that this is out of place. Perhaps this should be moved in to a `getSiacoinOutput` function (which doesn't exist at the moment).

File: modules/consensus/validtransaction.go [hunk 2]
```diff
@@ -25,8 +25,41 @@ var (
 	ErrUnfinishedFileContract     = errors.New("file contract window has not yet openend")
 	ErrUnrecognizedFileContractID = errors.New("cannot fetch storage proof segment for unknown file contract")
 	ErrWrongUnlockConditions      = errors.New("transaction contains incorrect unlock conditions")
+
+	// dont judge me
+	errSuccess = errors.New("please")
```
Comment (DavidVorick) [inline modules/consensus/validtransaction.go]: ah whoops

File: modules/consensus/applytransaction.go (hunk 1 above)
Comment (lukechampine) [inline modules/consensus/applytransaction.go]: are you going to leave the bare `panic` in there?

File: modules/consensus/validtransaction.go (hunk 2 above)
Comment (lukechampine) [inline modules/consensus/validtransaction.go]: kek

File: modules/consensus/validtransaction.go [hunk 3]
```diff
@@ -271,29 +495,32 @@ func (cs *ConsensusSet) validTransaction(t types.Transaction) error {
...
+				return err
+			}
 		}
+		return errSuccess
+	})
+	// Revert the siafund pool before checking the error.
+	cs.siafundPool = sfpOld
+	if err != errSuccess {
+		return modules.ConsensusChange{}, err
 	}
```
Comment (DavidVorick) [inline modules/consensus/validtransaction.go]: this is the code block that intentionally never commits the bolt tx.

File: modules/consensus/applytransaction.go (hunk 1 above)
Comment (DavidVorick) [inline modules/consensus/applytransaction.go]: oh, good catch.

Comment (DavidVorick) [inline modules/consensus/applytransaction.go]: I did fix this though github doesn't seem to think so for this comment window

**Pull request merged**